### PR TITLE
Encapsulation refactoring

### PR DESCRIPTION
**Pull request by rxin**

### CODE REVIEW

File: databricks/koalas/frame.py [hunk 1]
```diff
@@ -15,55 +15,62 @@
...
 from databricks.koalas.metadata import Metadata
 from databricks.koalas.missing.frame import _MissingPandasLikeDataFrame
 from databricks.koalas.selection import SparkDataFrameLocator
 
 
-class PandasLikeDataFrame(_Frame, _MissingPandasLikeDataFrame):
-    """
-    Methods that are relevant to dataframes.
-    """
+def default_session():
```
Comment (thunterdb) [inline databricks/koalas/frame.py]: in utils?

File: databricks/koalas/series.py [hunk 2]
```diff
@@ -15,46 +15,155 @@
...
+    return Series(scol, self._kdf, self._index_info)
+
+
+@decorator
+def _numpy_column_op(f, self, *args):
+    # PySpark does not support NumPy type out of the box. For now, we convert NumPy types
+    # into some primitive types understandable in PySpark.
+    new_args = []
+    for arg in args:
+        if isinstance(self.spark_type, LongType) and isinstance(arg, np.timedelta64):
```
Comment (thunterdb) [inline databricks/koalas/series.py]: is this working all the time? Put a TODO that this is a hack.

File: databricks/koalas/series.py [hunk 3]
```diff
@@ -15,46 +15,155 @@
...
+        self._kdf = kdf
+        self._index_info = index_info
+
+    # arithmetic operators
+    __neg__ = _column_op(spark.Column.__neg__)
+    __add__ = _column_op(spark.Column.__add__)
+
+    def __sub__(self, other):
+        # Note that timestamp subtraction casts arguments to integer. This is to mimic Pandas's
+        # behaviours. Pandas returns 'timedelta64[ns]' from 'datetime64[ns]'s subtraction.
```
Comment (thunterdb) [inline databricks/koalas/series.py]: behaviour

File: databricks/koalas/series.py [hunk 4]
```diff
@@ -15,46 +15,155 @@
...
+    new_args = []
+    for arg in args:
+        if isinstance(self.spark_type, LongType) and isinstance(arg, np.timedelta64):
+            new_args.append(float(arg / np.timedelta64(1, 's')))
+        else:
+            new_args.append(arg)
+    return _column_op(f)(self, *new_args)
+
+
+class Series(_Frame, _MissingPandasLikeSeries):
```
Comment (thunterdb) [inline databricks/koalas/series.py]: not Column anymore?

File: databricks/koalas/series.py [hunk 5]
```diff
@@ -256,20 +339,20 @@ def __dir__(self):
...
 
-def _unpack_scalar(df):
+def _unpack_scalar(sdf):
     """
     Takes a dataframe that is supposed to contain a single row with a single scalar value,
     and returns this value.
     """
-    l = df.head(2).collect()
-    assert len(l) == 1, (df, l)
+    l = sdf.head(2)
```
Comment (thunterdb) [inline databricks/koalas/series.py]: head(1) actually

File: databricks/koalas/tests/test_series_datetime.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import pandas as pd
+
+from databricks import koalas
+from databricks.koalas.testing.utils import ReusedSQLTestCase, TestUtils
+
+
+class SeriesDatetimeTest(ReusedSQLTestCase, TestUtils):
+
+    @property
+    def pdf1(self):
```
Comment (thunterdb) [inline databricks/koalas/tests/test_series_datetime.py]: are these tests coming from the pandas test suites?

File: databricks/koalas/frame.py [hunk 7]
```diff
@@ -15,55 +15,62 @@
...
-        if not hasattr(self, '_pandas_metadata') or self._pandas_metadata is None:
-            self._pandas_metadata = Metadata(column_fields=self.schema.fieldNames())
-        return self._pandas_metadata
+class DataFrame(_Frame, _MissingPandasLikeDataFrame):
 
-    @_metadata.setter
-    def _metadata(self, metadata):
-        self._pandas_metadata = metadata
+    @derived_from(pd.DataFrame)
+    @dispatch_on('data')
```
Comment (thunterdb) [inline databricks/koalas/frame.py]: i did not know this one...

File: databricks/koalas/frame.py [hunk 8]
```diff
@@ -15,55 +15,62 @@
...
-        * ``_pandas_metadata``: The metadata which stores column fields, and index fields and names.
-        """
-        self._pandas_metadata = None
 
-    @property
-    def _metadata(self):
-        if not hasattr(self, '_pandas_metadata') or self._pandas_metadata is None:
-            self._pandas_metadata = Metadata(column_fields=self.schema.fieldNames())
-        return self._pandas_metadata
+class DataFrame(_Frame, _MissingPandasLikeDataFrame):
```
Comment (thunterdb) [inline databricks/koalas/frame.py]: Please document all the hidden fields that you have, it is not easy to track with the dispatch (`_sdf`, `_metadata`). This is important for the reviewers.

File: databricks/koalas/series.py (hunk 4 above)
Comment (thunterdb) [reply databricks/koalas/series.py]: also, here as well, please document the internal fields.

**Comment by thunterdb** [comment]
Also, I feel very strongly that the most common Spark methods like alias, etc. should be added too very soon after that if you want to merge first.

**Comment by HyukjinKwon** [comment]
@rxin, I will take this over with addressing @thunterdb;s comments.

File: databricks/koalas/tests/test_series_datetime.py (hunk 6 above)
Comment (HyukjinKwon) [reply databricks/koalas/tests/test_series_datetime.py]: These tests were actually from the blog that will be published soon.

File: databricks/koalas/frame.py (hunk 7 above)
Comment (HyukjinKwon) [reply databricks/koalas/frame.py]: I didn't know too :D. `dispatch_on` is from third party library anyway. Actually I think we can get rid of `decorator` dependency at all later.

**Pull request closed without merging**